k40-whisperer: Add desktop item

sbaudlr opened the pull request:
###### Description of changes

This change adds a desktop item for the `k40-whisperer` package so that it appears in application launchers etc.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2475

piegamesde commented on pkgs/applications/misc/k40-whisperer/default.nix:
```diff
@@ -46,14 +48,29 @@ in stdenv.mkDerivation rec {
       --prefix PYTHONPATH : $out
   '';
 
+  desktopItem = makeDesktopItem {
```
Move this into a let binding please. I know you can put anything into `mkDerivation` and it will magically show up as shell variable in the build script, however this is horrible and should never have been possible in the first place and I really hope we'll get a sane mkDerivation one day …

piegamesde commented on pkgs/applications/misc/k40-whisperer/default.nix:
```diff
@@ -46,14 +48,29 @@ in stdenv.mkDerivation rec {
     ln -s ${udevRule} $out/lib/udev/rules.d/97-k40-whisperer.rules
+    ln -s "$desktopItem/share/applications/k40whisperer.desktop" "$out/share/applications/k40whisperer.desktop"
+
+    gm convert "$out/scorchworks.ico" "$out/share/icons/hicolor/128x128/apps/scorchworks.png"
```
I'm slightly in favor of renaming the icon into `k40-whisperer`

piegamesde commented on pkgs/applications/misc/k40-whisperer/default.nix:
```diff
@@ -46,14 +48,29 @@ in stdenv.mkDerivation rec {
+    mkdir -p "$out/share/applications"
 
     ln -s ${udevRule} $out/lib/udev/rules.d/97-k40-whisperer.rules
+    ln -s "$desktopItem/share/applications/k40whisperer.desktop" "$out/share/applications/k40whisperer.desktop"
```
Referencing variables before using this is weird, this confused me for a couple of minutes

sbaudlr commented:
I believe I have addressed your comments, although I'm not 100% sure I've understood https://github.com/NixOS/nixpkgs/pull/245437#discussion_r1277928031

SuperSandro2000 commented on pkgs/applications/misc/k40-whisperer/default.nix:
```diff
@@ -46,14 +58,21 @@ in stdenv.mkDerivation rec {
+    mkdir -p "$out/share/applications"
 
     ln -s ${udevRule} $out/lib/udev/rules.d/97-k40-whisperer.rules
+    ln -s "${desktopItem}/share/applications/k40-whisperer.desktop" "$out/share/applications/k40-whisperer.desktop"
```
Please use copyDesktopItems instead

SuperSandro2000 commented on pkgs/applications/misc/k40-whisperer/default.nix:
```diff
@@ -46,14 +58,21 @@ in stdenv.mkDerivation rec {
       --prefix PYTHONPATH : $out
   '';
 
+  desktopItems = [ desktopItem ];
```
This should be unused

piegamesde replied:
What do you mean? Should this line be removed?

SuperSandro2000 replied:
sorry, I can't remember

sbaudlr commented:
I've change it to use `copyDesktopItems`

piegamesde commented on pkgs/applications/misc/k40-whisperer/default.nix:
```diff
@@ -41,19 +54,29 @@ in stdenv.mkDerivation rec {
     mkdir -p $out/bin
     mkdir -p $out/lib/udev/rules.d
+    mkdir -p $out/share/icons/hicolor/128x128/apps
+    mkdir -p "$out/share/applications"
```
Is this still needed with `copyDesktopItems`?

sbaudlr replied:
Line 65 wasn't needed, I've removed it

SuperSandro2000 reviewed (approved):
the commits should probably be combined to no longer reflect the review process

sbaudlr committed:
k40-whisperer: Add desktop item

Co-authored-by: Pol Dellaiera <pol.dellaiera@protonmail.com>
